src/main/kotlin/uk/gov/justice/hmpps/offendersearch/services/SearchService.kt

```diff
@@ -0,0 +1,104 @@
+import uk.gov.justice.hmpps.offendersearch.dto.SearchDto
+import java.util.*
+
+@Service
+class SearchService @Autowired constructor(@param:Qualifier("elasticSearchClient") private val hlClient: RestHighLevelClient, private val mapper: ObjectMapper) {
+  companion object {
+    val log: Logger = LoggerFactory.getLogger(this::class.java)
+    private const val MAX_SEARCH_RESULTS = 100
+  }
+
+  fun performSearch(searchOptions: SearchDto): List<OffenderDetail> {
+    validateSearchForm(searchOptions)
+    val searchRequest = SearchRequest("offender")
+    val searchSourceBuilder = SearchSourceBuilder()
+    // Set the maximum search result size (the default would otherwise be 10)
+    searchSourceBuilder.size(MAX_SEARCH_RESULTS)
+    val matchingAllFieldsQuery = buildMatchWithAllProvidedParameters(searchOptions)
+    searchSourceBuilder.query(matchingAllFieldsQuery)
+    searchRequest.source(searchSourceBuilder)
+    val response = hlClient.search(searchRequest)
+    return getSearchResult(response)
+  }
+
+  protected fun buildMatchWithAllProvidedParameters(searchOptions: SearchDto): BoolQueryBuilder {
+    val matchingAllFieldsQuery = QueryBuilders
+        .boolQuery()
+    if (StringUtils.isNotBlank(searchOptions.surname)) {
+      matchingAllFieldsQuery.must(QueryBuilders
+          .matchQuery("surname", searchOptions.surname))
+    }
+    if (StringUtils.isNotBlank(searchOptions.firstName)) {
+      matchingAllFieldsQuery.must(QueryBuilders
+          .matchQuery("firstName", searchOptions.firstName))
+    }
+    if (searchOptions.dateOfBirth != null) {
+      matchingAllFieldsQuery.must(QueryBuilders
+          .matchQuery("dateOfBirth", searchOptions.dateOfBirth))
+    }
+    if (StringUtils.isNotBlank(searchOptions.crn)) {
+      matchingAllFieldsQuery.must(QueryBuilders
+          .matchQuery("otherIds.crn", searchOptions.crn))
+    }
+    if (StringUtils.isNotBlank(searchOptions.croNumber)) {
+      matchingAllFieldsQuery.must(QueryBuilders
+          .matchQuery("otherIds.croNumber", searchOptions.croNumber))
+    }
+    if (StringUtils.isNotBlank(searchOptions.pncNumber)) {
+      matchingAllFieldsQuery.must(QueryBuilders
+          .matchQuery("otherIds.pncNumber", searchOptions.pncNumber))
+    }
+    if (StringUtils.isNotBlank(searchOptions.nomsNumber)) {
+      matchingAllFieldsQuery.must(QueryBuilders
+          .matchQuery("otherIds.nomsNumber", searchOptions.nomsNumber))
+    }
+    return matchingAllFieldsQuery
+  }
+
+  private fun validateSearchForm(searchOptions: SearchDto) {
+    if (!searchOptions.isValid) {
+      log.warn("Invalid search  - no criteria provided")
+      throw BadRequestException("Invalid search  - please provide at least 1 search parameter")
+    }
+  }
+
+  private fun getSearchResult(response: SearchResponse): List<OffenderDetail> {
+    val searchHit = response.hits.hits
+    val offenderDetailList = ArrayList<OffenderDetail>()
+    if (searchHit.isNotEmpty()) {
+      Arrays.stream(searchHit)
```

Comment:
better to use kotlin streaming etc

Reply:
Agreed, something for a later date - didn't want to do too many changes